kubectl-fdb/cmd/cordon.go

```diff
@@ -116,54 +122,111 @@ kubectl fdb cordon -c cluster --node-selector machine=a,disk=fast
-				Selector: fields.OneTermEqualSelector("spec.nodeName", node),
-			})
-
+		pods, err := fetchPods(kubeClient, inputClusterName, namespace, node, clusterLabel)
 		if err != nil {
-			return err
+			operationFailed = true
+			errors = append(errors, fmt.Sprintf("Issue fetching Pods running on node: %s. Error: %s\n", node, err))
+			continue
 		}
 
-		for _, pod := range pods.Items {
-			// With the field selector above this shouldn't be required, but it's good to
-			// have a second check.
-			if pod.Spec.NodeName != node {
-				fmt.Printf("Pod: %s is not running on node %s will be ignored\n", pod.Name, node)
+		clusterNames := getClusterNames(cmd, inputClusterName, pods, clusterLabel)
+		for clusterName := range clusterNames {
+			cmd.Printf("Starting operation on %s, node: %s\n", clusterName, node)
+			cluster, err := loadCluster(kubeClient, namespace, clusterName)
+			if err != nil {
+				errors = append(errors, fmt.Sprintf("unable to load cluster: %s, skipping\n", clusterName))
+				operationFailed = true
 				continue
 			}
 
-			processGroup, ok := pod.Labels[cluster.GetProcessGroupIDLabel()]
-			if !ok {
-				fmt.Printf("could not fetch process group ID from Pod: %s\n", pod.Name)
-				continue
+			var processGroups []string
+			for _, pod := range pods.Items {
+				// With the field selector above this shouldn't be required, but it's good to
+				// have a second check.
+				if pod.Spec.NodeName != node {
+					cmd.Printf("Pod: %s is not running on node %s will be ignored\n", pod.Name, node)
+					continue
+				}
+
+				if cluster.ContainsPod(pod) {
+					processGroup, ok := pod.Labels[cluster.GetProcessGroupIDLabel()]
+					if !ok {
+						cmd.Printf("could not fetch process group ID from Pod: %s\n", pod.Name)
+						continue
+					}
+					processGroups = append(processGroups, processGroup)
+				}
+			}
+			err = replaceProcessGroups(kubeClient, cluster.Name, processGroups, namespace, withExclusion, wait, false, true, sleep)
+			if err != nil {
+				operationFailed = true
+				errors = append(errors, fmt.Sprintf("unable to cordon all Pods for cluster %s\n", cluster.Name))
 			}
-			processGroups = append(processGroups, processGroup)
 		}
 	}
+	if operationFailed {
+		return fmt.Errorf("following operation failed, please check and retry: \n, %s", errors)
+	}
+	return nil
+}
 
-	return replaceProcessGroups(kubeClient, cluster.Name, processGroups, namespace, withExclusion, wait, false, true, sleep)
+func fetchPods(kubeClient client.Client, inputClusterName string, namespace string, node string, clusterLabel string) (corev1.PodList, error) {
+	var pods corev1.PodList
+	var err error
+	if inputClusterName == "" {
+		err = kubeClient.List(ctx.Background(), &pods,
+			client.InNamespace(namespace),
+			client.HasLabels([]string{clusterLabel}),
+			client.MatchingFieldsSelector{
+				Selector: fields.OneTermEqualSelector("spec.nodeName", node),
+			})
+	} else {
+		cluster, error := loadCluster(kubeClient, namespace, inputClusterName)
```

Comment:
Now we are fetching the cluster multiple times (once for every node). This is probably okay for now, but we should create an issue to resolve this.

Reply:
https://github.com/FoundationDB/fdb-kubernetes-operator/issues/1496